lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -5982,11 +5990,8 @@ static void foldAutoDiffFunctionExtraction(
   if (!isInstructionTriviallyDead(source))
     return;
   source->eraseFromParent();
-  // Delete all worklist occurrences of `source` by setting them to nullptr.
-  // This is more efficient than APIs like `llvm::erase_if`.
-  for (auto &inst : context.getAutoDiffFunctionInsts())
-    if (inst == source)
-      inst = nullptr;
+  // Remove `source` from the worklist.
+  getAutoDiffFunctionInsts().remove(source);
```

Comment:
Removal from `SetVector<T>` is `O(n)`: http://llvm.org/docs/ProgrammersManual.html#llvm-adt-setvector-h. If you can, still, rewrite the element to `nullptr`, that would be good.  Otherwise, the best solution would be to revert it to `SmallVector` and use a separate `SmallPtrSet` to track pointers that have already been removed.

Reply:
Mutating `SetVector` elements doesn't seem to work:
```diff
@@ -5990,8 +5958,11 @@ void ADContext::foldAutoDiffFunctionExtraction(AutoDiffFunctionInst *source) {
   if (!isInstructionTriviallyDead(source))
     return;
   source->eraseFromParent();
-  // Remove `source` from the worklist.
-  getAutoDiffFunctionInsts().remove(source);
+  // Delete all worklist occurrences of `adfi` by setting them to nullptr.
+  // This is more efficient than APIs like `llvm::erase_if`.
+  for (auto &inst : getAutoDiffFunctionInsts())
+    if (inst == source)
+      inst = nullptr;
 }
```

```
/Users/danielzheng/swift-merge/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:5965:12: error: cannot assign to variable 'inst' with const-qualified type 'swift::AutoDiffFunctionInst *const &'
      inst = nullptr;
      ~~~~ ^
/Users/danielzheng/swift-merge/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:5963:14: note: variable 'inst' declared const here
  for (auto &inst : getAutoDiffFunctionInsts())
       ~~~~~~^~~~~~
1 error generated.
```

I'll follow your suggestion of `SmallVector` + `SmallPtrSet`!

Reply:
Done in https://github.com/apple/swift/pull/24946/commits/eba1c455d12732d7892430cff19c6dc0b8f691e3.

```diff
+  /// The set of `autodiff_function` instructions that have been processed.
+  /// Used to avoid reprocessing invalidated instructions.
+  SmallPtrSet<AutoDiffFunctionInst *, 32> processedAutoDiffFunctionInsts;

@@ -5982,11 +5955,8 @@ static void foldAutoDiffFunctionExtraction(
   if (!isInstructionTriviallyDead(source))
     return;
   source->eraseFromParent();
-  // Delete all worklist occurrences of `source` by setting them to nullptr.
-  // This is more efficient than APIs like `llvm::erase_if`.
-  for (auto &inst : context.getAutoDiffFunctionInsts())
-    if (inst == source)
-      inst = nullptr;
+  // Mark `source` as processed so that it won't be reprocessed after deletion.
+  processedAutoDiffFunctionInsts.insert(source);

@@ -6080,9 +6047,8 @@ void Differentiation::run() {
   while (!context.getAutoDiffFunctionInsts().empty()) {
     auto *adfi = context.getAutoDiffFunctionInsts().back();
     context.getAutoDiffFunctionInsts().pop_back();
-    // Skip instructions that have been set to nullptr by
-    // `processAutoDiffFunctionInst`.
-    if (!adfi) continue;
+    // Skip instructions that have been already been processed.
+    if (context.getProcessedAutoDiffFunctionInsts().count(adfi)) continue;
     errorOccurred |= context.processAutoDiffFunctionInst(adfi);
   }
```